virtual_rainforest/core/config.py

```diff
@@ -0,0 +1,362 @@
+        if module_name in SCHEMA_REGISTRY:
+            log_and_raise(
+                f"The module schema {module_name} is used multiple times, this "
+                f"shouldn't be the case!",
+                ValueError,
+            )
+        else:
+            # Check that this is a valid schema
+            try:
+                jsonschema.Draft202012Validator.check_schema(func())
+            except jsonschema.exceptions.SchemaError:
+                log_and_raise(
+                    f"Module schema {module_name} not valid JSON!",
+                    OSError,
+                )
+            # If it is valid then add it to the registry
+            SCHEMA_REGISTRY[module_name] = func()
+
+        return func
+
+    return wrap
+
+
+# Dictionary to store validated config
+COMPLETE_CONFIG: dict = {}
+
+
+def check_dict_leaves(d1: dict, d2: dict, conflicts: list = [], path: list = []):
+    """Recursively checks if leaves are repeated between two nested dictionaries.
+
+    Args:
+        d1: First nested dictionary to compare
+        d2: Second nested dictionary to compare
+        path: List describing recursive path through the nested dictionary
+            conflicts: List of variables that are defined in multiple places
+
+    Returns:
+        conflicts: List of variables that are defined in multiple places
+    """
+
+    for key in d2:
+        if key in d1:
+            if isinstance(d1[key], dict) and isinstance(d2[key], dict):
+                check_dict_leaves(d1[key], d2[key], conflicts, path + [str(key)])
+            else:
+                conflicts.append("%s" % ".".join(path + [str(key)]))
+
+    return conflicts
+
+
+def check_outfile(output_folder: str, out_file_name: str) -> None:
+    """Check that final output file is not already in the output folder.
+
+    Args:
+        output_folder: Path to a folder to output the outputted complete configuration
+            file to
+        out_file_name: The name to save the outputted complete configuration file under
+    """
+
+    # Throw critical error if combined output file already exists
+    for file in Path(output_folder).iterdir():
+        if file.name == f"{out_file_name}.toml":
+            log_and_raise(
+                f"A config file in the specified configuration folder already makes use"
+                f" of the specified output file name ({out_file_name}.toml), this file "
+                f"should either be renamed or deleted!",
+                OSError,
+            )
+
+    return None
+
+
+def collect_files(cfg_paths: list[str]) -> list[Path]:
+    """Collect all toml files from a user specified list of files and directories.
+
+    Args:
+        cfg_paths: A path or a set of paths that point to either configuration files, or
+            folders containing configuration files
+    """
+
+    # Preallocate file list
+    files = []
+    not_found = []  # Stores all invalid paths
+    empty_fold = []  # Stores all empty toml folders
```

Comment:
Not totally sure if these lists will store `Path`, `str` or a mixture of both. Maybe add some type annotations here?